src/mds/SnapRealm.h

```diff
@@ -147,7 +147,7 @@ struct SnapRealm {
   }
   void remove_cap(client_t client, Capability *cap) {
     cap->item_snaprealm_caps.remove_myself();
-    if (client_caps[client]->empty()) {
+    if (client_caps.count(client) && client_caps[client]->empty()) {
```

Comment:
this is an anti-pattern to use STL. as you are unnecessarily looking up in the `map<>` twice (actually for four times!).

i'd suggest something like

```c++
if (auto found = client_caps.find(client); 
    found != client_caps.end() && found->empty()) {
  delete *found;
  client_caps.erase(found);
}
```